### DISCUSSION TITLE
What is the better way to send request by programming in AI Feature

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-theia/theia#14336] [Q&A]
Hi,
    When some event triggers I want to send a request in AI feature of theia,  what is the better way to send a request by programming?
 I have seen this.chatService.sendRequest, is it elegant to invoke this method directly?

**Comment by wss29** [comment]
currently I use follow code to send request by programming in my class
```
    public async sendMessage(message: string): Promise<void> {
        const chatSession = this.chatService.getSessions().find(o => o.isActive);
        if (chatSession) {
            const chatRequest: ChatRequest = {
                text: message,
            };

            const requestProgress = await this.chatService.sendRequest(chatSession.id, chatRequest);
            requestProgress?.responseCompleted.then(responseModel => {
                if (responseModel.isError) {
                    this.messageService.error(responseModel.errorObject?.message ?? 'An error occurred druring chat service invocation.');
                }
            });
            if (!requestProgress) {
                this.messageService.error(`Was not able to send request "${chatRequest.text}" to session ${chatSession.id}`);
                return;
            }
        }
    }
```

**Reply by sdirix** [reply]
Generally speaking the code is fine. It's the [same code we use in the chat widget](https://github.com/eclipse-theia/theia/blob/c1e4279cf85aff0b4fdfccbe605fc391aa599e0a/packages/ai-chat-ui/src/browser/chat-view-widget.tsx#L156-L169) provided in Theia.

> When some event triggers I want to send a request in AI feature of theia, what is the better way to send a request by programming?
I have seen this.chatService.sendRequest, is it elegant to invoke this method directly?

If you like you can go through the chat service. If you don't really need a chat but prefer a more direct interaction with the underlying framework, then you can also interact with agents directly, [see for example the terminal integration](https://github.com/eclipse-theia/theia/blob/c1e4279cf85aff0b4fdfccbe605fc391aa599e0a/packages/ai-terminal/src/browser/ai-terminal-contribution.ts#L157).
